File: src/workspaceforge/scan/stack-detector.ts
import path from "node:path";
import { promises as fs } from "node:fs";
import { fileExists, listDir } from "../util/fs-utils.js";

export type PackageManager = "npm" | "pnpm" | "yarn" | "pip" | "poetry" | "cargo" | "go" | "bundler" | "unknown";

export type PrimaryLanguage =
  | "typescript"
  | "javascript"
  | "python"
  | "rust"
  | "go"
  | "ruby"
  | "unknown";

export interface MonorepoInfo {
  isMonorepo: boolean;
  packageCount: number;
  packagePaths: string[];
}

export interface StackDetectionResult {
  language: PrimaryLanguage;
  packageManager: PackageManager;
  monorepo: MonorepoInfo;
  /** Raw signals found, for status/debug output -- e.g. ["package-lock.json", "tsconfig.json"]. */
  signals: string[];
}

interface PackageJsonShape {
  workspaces?: string[] | { packages?: string[] };
  devDependencies?: Record<string, string>;
  dependencies?: Record<string, string>;
}

/**
 * WF01 -- stack and package-manager detection. Real filesystem checks, no
 * network calls. Covers, at minimum, npm/pnpm/yarn + TypeScript/JavaScript,
 * plus Python (pip/poetry), with lighter-weight signals for cargo, go
 * modules, and bundler/Gemfile per the broader workspace-surface scan table.
 */
export async function detectStack(repoPath: string): Promise<StackDetectionResult> {
  const signals: string[] = [];

  const [
    hasPackageJson,
    hasPackageLock,
    hasPnpmLock,
    hasYarnLock,
    hasPnpmWorkspaceYaml,
    hasTsconfig,
    hasRequirementsTxt,
    hasPyprojectToml,
    hasPipfile,
    hasCargoToml,
    hasGoMod,
    hasGemfile,
  ] = await Promise.all([
    fileExists(path.join(repoPath, "package.json")),
    fileExists(path.join(repoPath, "package-lock.json")),
    fileExists(path.join(repoPath, "pnpm-lock.yaml")),
    fileExists(path.join(repoPath, "yarn.lock")),
    fileExists(path.join(repoPath, "pnpm-workspace.yaml")),
    fileExists(path.join(repoPath, "tsconfig.json")),
    fileExists(path.join(repoPath, "requirements.txt")),
    fileExists(path.join(repoPath, "pyproject.toml")),
    fileExists(path.join(repoPath, "Pipfile")),
    fileExists(path.join(repoPath, "Cargo.toml")),
    fileExists(path.join(repoPath, "go.mod")),
    fileExists(path.join(repoPath, "Gemfile")),
  ]);

  let packageManager: PackageManager = "unknown";
  let language: PrimaryLanguage = "unknown";

  if (hasPackageJson) {
    signals.push("package.json");
    if (hasPnpmLock || hasPnpmWorkspaceYaml) {
      packageManager = "pnpm";
      signals.push(hasPnpmLock ? "pnpm-lock.yaml" : "pnpm-workspace.yaml");
    } else if (hasYarnLock) {
      packageManager = "yarn";
      signals.push("yarn.lock");
    } else if (hasPackageLock) {
      packageManager = "npm";
      signals.push("package-lock.json");
    } else {
      // package.json present with no lockfile detected yet -- default to npm,
      // the ecosystem default, rather than leaving it unknown.
      packageManager = "npm";
    }

    if (hasTsconfig) {
      language = "typescript";
      signals.push("tsconfig.json");
    } else {
      language = "javascript";
    }
  } else if (hasPyprojectToml) {
    signals.push("pyproject.toml");
    packageManager = "poetry";
    language = "python";
  } else if (hasPipfile) {
    signals.push("Pipfile");
    packageManager = "pip";
    language = "python";
  } else if (hasRequirementsTxt) {
    signals.push("requirements.txt");
    packageManager = "pip";
    language = "python";
  } else if (hasCargoToml) {
    signals.push("Cargo.toml");
    packageManager = "cargo";
    language = "rust";
  } else if (hasGoMod) {
    signals.push("go.mod");
    packageManager = "go";
    language = "go";
  } else if (hasGemfile) {
    signals.push("Gemfile");
    packageManager = "bundler";
    language = "ruby";
  }

  const monorepo = await detectMonorepo(repoPath, packageManager, hasPackageJson);

  return { language, packageManager, monorepo, signals };
}

async function detectMonorepo(
  repoPath: string,
  packageManager: PackageManager,
  hasPackageJson: boolean,
): Promise<MonorepoInfo> {
  if (!hasPackageJson) {
    return { isMonorepo: false, packageCount: 0, packagePaths: [] };
  }

  let workspaceGlobs: string[] = [];

  if (packageManager === "pnpm") {
    const pnpmWorkspacePath = path.join(repoPath, "pnpm-workspace.yaml");
    if (await fileExists(pnpmWorkspacePath)) {
      const raw = await fs.readFile(pnpmWorkspacePath, "utf-8");
      workspaceGlobs = parseSimpleYamlList(raw);
    }
  }

  if (workspaceGlobs.length === 0) {
    const packageJsonPath = path.join(repoPath, "package.json");
    try {
      const raw = await fs.readFile(packageJsonPath, "utf-8");
      const parsed = JSON.parse(raw) as PackageJsonShape;
      if (Array.isArray(parsed.workspaces)) {
        workspaceGlobs = parsed.workspaces;
      } else if (parsed.workspaces?.packages) {
        workspaceGlobs = parsed.workspaces.packages;
      }
    } catch {
      // Malformed package.json -- treat as no workspace info rather than throw.
    }
  }

  if (workspaceGlobs.length === 0) {
    return { isMonorepo: false, packageCount: 0, packagePaths: [] };
  }

  const packagePaths = await resolveWorkspacePackages(repoPath, workspaceGlobs);
  return {
    isMonorepo: packagePaths.length > 0,
    packageCount: packagePaths.length,
    packagePaths,
  };
}

/**
 * Minimal glob resolution for the common `dir/*` workspace pattern plus exact
 * directory entries. This intentionally does not pull in a glob dependency --
 * v0.1's monorepo detection covers the overwhelmingly common shapes
 * (`packages/*`, `apps/*`, explicit paths) without adding a runtime dep.
 */
async function resolveWorkspacePackages(repoPath: string, globs: string[]): Promise<string[]> {
  const resolved = new Set<string>();

  for (const glob of globs) {
    if (glob.startsWith("!")) {
      continue; // negation patterns -- not supported in v0.1, skip rather than mis-include.
    }
    if (glob.endsWith("/*")) {
      const baseDir = glob.slice(0, -2);
      const absoluteBase = path.join(repoPath, baseDir);
      const entries = await listDir(absoluteBase);
      for (const entry of entries) {
        const candidate = path.join(absoluteBase, entry);
        if (await fileExists(path.join(candidate, "package.json"))) {
          resolved.add(path.relative(repoPath, candidate));
        }
      }
    } else {
      const absolute = path.join(repoPath, glob);
      if (await fileExists(path.join(absolute, "package.json"))) {
        resolved.add(path.relative(repoPath, absolute));
      }
    }
  }

  return Array.from(resolved).sort();
}

/** Extremely small YAML list parser, just for pnpm-workspace.yaml's `packages:` block. */
function parseSimpleYamlList(raw: string): string[] {
  const lines = raw.split("\n");
  const items: string[] = [];
  let inPackagesBlock = false;
  for (const line of lines) {
    if (/^packages\s*:/.test(line.trim())) {
      inPackagesBlock = true;
      continue;
    }
    if (inPackagesBlock) {
      const match = line.match(/^\s*-\s*['"]?([^'"#]+)['"]?/);
      if (match?.[1]) {
        items.push(match[1].trim());
      } else if (line.trim().length > 0 && !line.startsWith(" ") && !line.startsWith("-")) {
        break; // dedented past the packages block
      }
    }
  }
  return items;
}

File: python/src/agenticworkspace/scan/stack_detector.py
"""
Stack and package-manager detection. Real filesystem checks, no network
calls. Covers, at minimum, npm/pnpm/yarn + TypeScript/JavaScript, plus Python
(pip/poetry), with lighter-weight signals for cargo, go modules, and
bundler/Gemfile. Ported from src/agenticworkspace/scan/stack-detector.ts.
"""
from __future__ import annotations

import json
import os
import re
from dataclasses import dataclass, field
from typing import List, Set

from ..util.fs_utils import file_exists, list_dir

PackageManager = str  # "npm" | "pnpm" | "yarn" | "pip" | "poetry" | "cargo" | "go" | "bundler" | "unknown"
PrimaryLanguage = str  # "typescript" | "javascript" | "python" | "rust" | "go" | "ruby" | "unknown"


@dataclass
class MonorepoInfo:
    is_monorepo: bool
    package_count: int
    package_paths: List[str]


@dataclass
class StackDetectionResult:
    language: PrimaryLanguage
    package_manager: PackageManager
    monorepo: MonorepoInfo
    # Raw signals found, for status/debug output -- e.g. ["package-lock.json", "tsconfig.json"].
    signals: List[str] = field(default_factory=list)


def detect_stack(repo_path: str) -> StackDetectionResult:
    signals: List[str] = []

    has_package_json = file_exists(os.path.join(repo_path, "package.json"))
    has_package_lock = file_exists(os.path.join(repo_path, "package-lock.json"))
    has_pnpm_lock = file_exists(os.path.join(repo_path, "pnpm-lock.yaml"))
    has_yarn_lock = file_exists(os.path.join(repo_path, "yarn.lock"))
    has_pnpm_workspace_yaml = file_exists(os.path.join(repo_path, "pnpm-workspace.yaml"))
    has_tsconfig = file_exists(os.path.join(repo_path, "tsconfig.json"))
    has_requirements_txt = file_exists(os.path.join(repo_path, "requirements.txt"))
    has_pyproject_toml = file_exists(os.path.join(repo_path, "pyproject.toml"))
    has_pipfile = file_exists(os.path.join(repo_path, "Pipfile"))
    has_cargo_toml = file_exists(os.path.join(repo_path, "Cargo.toml"))
    has_go_mod = file_exists(os.path.join(repo_path, "go.mod"))
    has_gemfile = file_exists(os.path.join(repo_path, "Gemfile"))

    package_manager: PackageManager = "unknown"
    language: PrimaryLanguage = "unknown"

    if has_package_json:
        signals.append("package.json")
        if has_pnpm_lock or has_pnpm_workspace_yaml:
            package_manager = "pnpm"
            signals.append("pnpm-lock.yaml" if has_pnpm_lock else "pnpm-workspace.yaml")
        elif has_yarn_lock:
            package_manager = "yarn"
            signals.append("yarn.lock")
        elif has_package_lock:
            package_manager = "npm"
            signals.append("package-lock.json")
        else:
            # package.json present with no lockfile detected yet -- default to npm,
            # the ecosystem default, rather than leaving it unknown.
            package_manager = "npm"

        if has_tsconfig:
            language = "typescript"
            signals.append("tsconfig.json")
        else:
            language = "javascript"
    elif has_pyproject_toml:
        signals.append("pyproject.toml")
        package_manager = "poetry"
        language = "python"
    elif has_pipfile:
        signals.append("Pipfile")
        package_manager = "pip"
        language = "python"
    elif has_requirements_txt:
        signals.append("requirements.txt")
        package_manager = "pip"
        language = "python"
    elif has_cargo_toml:
        signals.append("Cargo.toml")
        package_manager = "cargo"
        language = "rust"
    elif has_go_mod:
        signals.append("go.mod")
        package_manager = "go"
        language = "go"
    elif has_gemfile:
        signals.append("Gemfile")
        package_manager = "bundler"
        language = "ruby"

    monorepo = _detect_monorepo(repo_path, package_manager, has_package_json)

    return StackDetectionResult(language=language, package_manager=package_manager, monorepo=monorepo, signals=signals)


def _detect_monorepo(repo_path: str, package_manager: PackageManager, has_package_json: bool) -> MonorepoInfo:
    if not has_package_json:
        return MonorepoInfo(is_monorepo=False, package_count=0, package_paths=[])

    workspace_globs: List[str] = []

    if package_manager == "pnpm":
        pnpm_workspace_path = os.path.join(repo_path, "pnpm-workspace.yaml")
        if file_exists(pnpm_workspace_path):
            with open(pnpm_workspace_path, "r", encoding="utf-8") as handle:
                raw = handle.read()
            workspace_globs = _parse_simple_yaml_list(raw)

    if len(workspace_globs) == 0:
        package_json_path = os.path.join(repo_path, "package.json")
        try:
            with open(package_json_path, "r", encoding="utf-8") as handle:
                raw = handle.read()
            parsed = json.loads(raw)
            workspaces = parsed.get("workspaces") if isinstance(parsed, dict) else None
            if isinstance(workspaces, list):
                workspace_globs = workspaces
            elif isinstance(workspaces, dict) and isinstance(workspaces.get("packages"), list):
                workspace_globs = workspaces["packages"]
        except (OSError, ValueError):
            # Malformed package.json -- treat as no workspace info rather than raise.
            pass

    if len(workspace_globs) == 0:
        return MonorepoInfo(is_monorepo=False, package_count=0, package_paths=[])

    package_paths = _resolve_workspace_packages(repo_path, workspace_globs)
    return MonorepoInfo(
        is_monorepo=len(package_paths) > 0,
        package_count=len(package_paths),
        package_paths=package_paths,
    )


def _resolve_workspace_packages(repo_path: str, globs: List[str]) -> List[str]:
    """
    Minimal glob resolution for the common `dir/*` workspace pattern plus exact
    directory entries. This intentionally does not pull in a glob dependency --
    v0.1's monorepo detection covers the overwhelmingly common shapes
    (`packages/*`, `apps/*`, explicit paths) without adding a runtime dep.
    """
    resolved: Set[str] = set()

    for glob in globs:
        if glob.startswith("!"):
            continue  # negation patterns -- not supported in v0.1, skip rather than mis-include.
        if glob.endswith("/*"):
            base_dir = glob[:-2]
            absolute_base = os.path.join(repo_path, base_dir)
            entries = list_dir(absolute_base)
            for entry in entries:
                candidate = os.path.join(absolute_base, entry)
                if file_exists(os.path.join(candidate, "package.json")):
                    _add_if_within_repo(resolved, repo_path, candidate)
        else:
            absolute = os.path.join(repo_path, glob)
            if file_exists(os.path.join(absolute, "package.json")):
                _add_if_within_repo(resolved, repo_path, absolute)

    return sorted(resolved)


def _add_if_within_repo(resolved: Set[str], repo_path: str, candidate: str) -> None:
    """
    The scanned repo's own package.json/pnpm-workspace.yaml declares these
    workspace globs, so a malicious or untrusted repo (this tool's whole
    purpose is onboarding arbitrary repos) could declare a glob like
    "../sibling-project" to make AgenticWorkspace resolve and reference a path
    outside the repo it was asked to scan. Confine every resolved candidate to
    repo_path; drop anything that escapes it rather than silently including it.
    """
    relative = os.path.relpath(candidate, repo_path)
    if relative.startswith("..") or os.path.isabs(relative):
        return
    resolved.add(relative)


_PACKAGES_KEY_RE = re.compile(r"^packages\s*:")
_YAML_LIST_ITEM_RE = re.compile(r"^\s*-\s*['\"]?([^'\"#]+)['\"]?")


def _parse_simple_yaml_list(raw: str) -> List[str]:
    """Extremely small YAML list parser, just for pnpm-workspace.yaml's `packages:` block."""
    lines = raw.split("\n")
    items: List[str] = []
    in_packages_block = False
    for line in lines:
        if _PACKAGES_KEY_RE.match(line.strip()):
            in_packages_block = True
            continue
        if in_packages_block:
            match = _YAML_LIST_ITEM_RE.match(line)
            if match and match.group(1):
                items.append(match.group(1).strip())
            elif len(line.strip()) > 0 and not line.startswith(" ") and not line.startswith("-"):
                break  # dedented past the packages block
    return items
